packages/dds/tree/src/tableSchema.ts

```diff
@@ -745,96 +723,178 @@ export namespace System_TableSchema {
 			}
 
 			public removeColumns(
-				columns: readonly string[] | readonly ColumnValueType[],
+				indexOrColumns: number | readonly string[] | readonly ColumnValueType[],
+				end: number | undefined = undefined,
 			): ColumnValueType[] {
+				if (typeof indexOrColumns === "number") {
+					return this.removeRangeOfColumns(indexOrColumns, end);
+				}
+
+				const columnsToRemove = indexOrColumns;
+
 				// If there are no columns to remove, do nothing
-				if (columns.length === 0) {
+				if (columnsToRemove.length === 0) {
 					return [];
 				}
 
 				// If there is only one column to remove, remove it (and don't incur cost of transaction)
-				if (columns.length === 1) {
-					const removedColumn = this.removeColumn(columns[0] ?? oob());
+				if (columnsToRemove.length === 1) {
+					const columnOrIdToRemove = columnsToRemove[0] ?? oob();
+					const index = this._getColumnIndex(columnOrIdToRemove);
+
+					if (index === undefined) {
+						const columnId = this._getColumnId(columnOrIdToRemove);
+						throw new UsageError(
+							`Specified column with ID "${columnId}" does not exist in the table.`,
+						);
+					}
+					const removedColumn = this.columns[index] as ColumnValueType;
+					this.columns.removeAt(index);
 					return [removedColumn];
 				}
 
 				// If there are multiple columns to remove, remove them in a transaction.
-				const removedColumns: ColumnValueType[] = [];
+
+				// To reduce the number of operations in the transaction, we will build up a series of
+				// contiguous ranges to be removed from the input list and remove each range as a single operation.
+				const columnIndicesToRemove: number[] = [];
+				const columnNodesToRemove: ColumnValueType[] = [];
+				for (const column of columnsToRemove) {
+					const columnIndex = this._getColumnIndex(column);
+					if (columnIndex === undefined) {
+						const columnId = this._getColumnId(column);
+						throw new UsageError(
+							`Specified column with ID "${columnId}" does not exist in the table.`,
+						);
+					}
+					columnIndicesToRemove.push(columnIndex);
+					columnNodesToRemove.push(this.columns[columnIndex] as ColumnValueType);
+				}
+
+				const rangesToRemove = Table._groupIndicesIntoRanges(columnIndicesToRemove);
+
 				Tree.runTransaction(this, () => {
 					// Note, throwing an error within a transaction will abort the entire transaction.
-					// So if we throw an error here for any row, no columns will be removed.
-					for (const columnToRemove of columns) {
-						const removedRow = this.removeColumn(columnToRemove);
-						removedColumns.push(removedRow);
+					// So if we throw an error here for any column, no columns will be removed.
+					for (const range of rangesToRemove) {
+						this.removeRangeOfColumns(range.start, range.end);
```

Comment:
TODO: this is wrong (for both columns and rows). Each iteration we need to adjust the remaining ranges to account for the previously removed items